File: api/app/utils/conversation_state.py
import time

conversation_states = {}


def set_conversation_status(discord_id: str, state: dict):
    conversation_states[discord_id] = {**state, "expires_at": time.time() + 300}


def get_conversation_status(discord_id: str) -> dict | None:
    state = conversation_states.get(discord_id)
    if not state:
        return None
    if time.time() > state["expires_at"]:
        del conversation_states[discord_id]
        return None
    return state


def clear_conversation_status(discord_id: str):
    conversation_states.pop(discord_id, None)

File: api/src/utils/conversationState.ts
interface ConversationState {
  state: string;
  expires_at: number;
  [key: string]: unknown;
}

const conversationStates = new Map<string, ConversationState>();

export function setConversationStatus(discordId: string, state: Record<string, unknown>) {
  conversationStates.set(discordId, { ...state, state: state["state"] as string, expires_at: Date.now() + 300_000 });
}

export function getConversationStatus(discordId: string): ConversationState | null {
  const state = conversationStates.get(discordId);
  if (!state) return null;
  if (Date.now() > state.expires_at) {
    conversationStates.delete(discordId);
    return null;
  }
  return state;
}

export function clearConversationStatus(discordId: string) {
  conversationStates.delete(discordId);
}
